src/main/java/com/wultra/app/enrollmentserver/impl/service/converter/MobileTokenConverter.java

```
@@ -0,0 +1,139 @@
+                    }
+                }
+            }
+
+            operation.setFormData(formData);
+            return operation;
+        } catch (JsonProcessingException e) {
+            logger.debug("Unable to parse JSON with operation template parameters: {}", e.getMessage());
+            logger.debug("Exception detail", e);
+            throw new MobileTokenConfigurationException("ERR_CONFIG", "Invalid JSON structure for the configuration: " + e.getMessage());
+        }
+    }
+
+    private Attribute buildAttribute(OperationTemplateParam templateParam, Map<String, String> params) {
+        final String type = templateParam.getType();
+        final String id = templateParam.getId();
+        final String text = templateParam.getText();
+        switch (type) {
+            case "AMOUNT": {
+                final String amountKey = templateParam.getParams().get("amount");
+                final String currencyKey = templateParam.getParams().get("currency");
+                final String amount = params.get(amountKey);
+                final String currency = params.get(currencyKey);
+                try {
+                    final BigDecimal amountValue = new BigDecimal(amount);
+                    return new AmountAttribute(id, text, amountValue, currency, amount, currency);
+                } catch (NumberFormatException ex) {
+                    logger.warn("Invalid number format: {}, skipping the AMOUNT attribute!", amount);
+                    return null;
+                }
+            }
+            case "HEADING": {
+                return new HeadingAttribute(id, text);
+            }
+            case "NOTE": {
+                final String noteKey = templateParam.getParams().get("note");
+                final String note = params.get(noteKey);
+                return new NoteAttribute(id, text, note);
+            }
+            case "KEY_VALUE":
+            default: { // attempt fallback to key-value type
```

I would prefer to throw an error for an unknown type to avoid surprises (e.g. null values).

---

We generally handle this with the fallbacks in the mobile app too. I will keep the fallback and add error logging instead.